src/CVRF/cvrf_eval.c

```diff
@@ -0,0 +1,470 @@
+	__attribute__nonnull__(export_file);
+
+	struct cvrf_session *session = cvrf_session_new_from_source_model(import_source);
+	cvrf_session_set_os_name(session, os_name);
+	cvrf_session_set_results_file(session, export_file);
+
+	if (find_all_cvrf_product_ids_from_cpe(session) != 0) {
+		cvrf_session_free(session);
+		return -1;
+	}
+	cvrf_session_construct_definition_model(session);
+
+	xmlDocPtr doc = xmlNewDoc(BAD_CAST "1.0");
+	if (doc == NULL) {
+		oscap_setxmlerr(xmlGetLastError());
+	}
+	xmlNode *model_node = cvrf_model_results_to_dom(session);
+	xmlDocSetRootElement(doc, model_node);
+
+	struct oscap_source *source = oscap_source_new_from_xmlDoc(doc, export_file);
+	int ret = oscap_source_save_as(source, NULL);
+	oscap_source_free(source);
+	cvrf_session_free(session);
+	return ret;
+}
+
+struct oscap_source *cvrf_index_get_results_source(struct oscap_source *import_source, const char *os_name) {
+	struct cvrf_session *session = cvrf_session_new_from_source_index(import_source);
+	cvrf_session_set_os_name(session, os_name);
+
+	xmlDocPtr doc = xmlNewDoc(BAD_CAST "1.0");
+	if (doc == NULL) {
+		oscap_setxmlerr(xmlGetLastError());
+	}
+	xmlNode *index_node = xmlNewNode(NULL, BAD_CAST "Index");
+	xmlDocSetRootElement(doc, index_node);
+
+	struct cvrf_model_iterator *it = cvrf_index_get_models(session->index);
+	while (cvrf_model_iterator_has_more(it)) {
+		struct cvrf_model *model = cvrf_model_iterator_next(it);
+		session->model = model;
+		find_all_cvrf_product_ids_from_cpe(session);
+		cvrf_session_construct_definition_model(session);
+
+		xmlNode *model_node = cvrf_model_results_to_dom(session);
+		xmlAddChild(index_node, model_node);
+	}
+	cvrf_model_iterator_free(it);
+
+	struct oscap_source *source = oscap_source_new_from_xmlDoc(doc, NULL);
+	cvrf_session_free(session);
+	return source;
+}
+
+static const char *get_rpm_name_from_cvrf_product_id(struct cvrf_session *session, const char *product_id) {
+	const char *rpm_name = NULL;
+	struct cvrf_product_tree *tree = cvrf_model_get_product_tree(session->model);
+
+	struct oscap_iterator *branches = cvrf_product_tree_get_branches(tree);
+	while (oscap_iterator_has_more(branches)) {
+		struct cvrf_branch *branch = oscap_iterator_next(branches);
+		if (cvrf_branch_get_branch_type(branch) == CVRF_BRANCH_PRODUCT_VERSION) {
+			struct cvrf_product_name *full_name = cvrf_branch_get_product_name(branch);
+
+			if (oscap_str_endswith(product_id, cvrf_product_name_get_product_id(full_name))) {
+				rpm_name = cvrf_product_name_get_cpe(full_name);
+				break;
+			}
+		}
+	}
+	oscap_iterator_free(branches);
+	return rpm_name;
+}
+
+static struct cvrf_rpm_attributes *parse_rpm_attributes_from_cvrf_product_id(struct cvrf_session *session,
+		const char *product_id) {
+
+	struct cvrf_rpm_attributes *attributes = cvrf_rpm_attributes_new();
+	attributes->full_package_name = strdup(get_rpm_name_from_cvrf_product_id(session, product_id));
+	char *package = strdup(strchr(product_id, ':')+1);
```

Comment:
What if there is no semicolon in the string `product_id`?  If there's no semicolon in the string, function `strchr` will return NULL, it will add 1 to NULL, so it will do `strdup(NULL+1)`, which I'm afraid will crash horribly.